src/overlays/actors/ovl_Boss_Hakugin/z_boss_hakugin.c

```diff
@@ -46,342 +113,3711 @@ ActorInit Boss_Hakugin_InitVars = {
+        velocity.y = (this->timer - 30) * (1 / 70.0f);
+    } else {
+        velocity.y = 0.1f;
+    }
+
+    velocity.x = Rand_CenteredFloat(2.0f) * velocity.y;
+    velocity.z = Rand_CenteredFloat(2.0f) * velocity.y;
+
+    pos.x = ((velocity.x >= 0.0f ? 1.0f : -1.0f) * (Rand_ZeroFloat(20.0f) + 5.0f) * 4.0f) + this->iceCollider.dim.pos.x;
+    pos.z = ((velocity.z >= 0.0f ? 1.0f : -1.0f) * (Rand_ZeroFloat(20.0f) + 5.0f) * 4.0f) + this->iceCollider.dim.pos.z;
+
+    if (steamFollowsMeltingIce) {
+        pos.y = this->iceCollider.dim.pos.y + (this->iceCollider.dim.height * velocity.y);
+    } else {
+        pos.y = (Rand_ZeroFloat(this->iceCollider.dim.height) * 0.8f) + this->iceCollider.dim.pos.y;
+    }
+
+    velocity.y += 0.8f;
+    EffectSsIceSmoke_Spawn(play, &pos, &velocity, &gZeroVec3f, 600);
+}
+
+/**
+ * Returns true if Goht should stop running and either wait for the player to get in view (to shoot lightning at them)
+ * or wait 150 frames (after which, Goht will turn around and charge them).
+ */
+s32 BossHakugin_ShouldWait(BossHakugin* this, PlayState* play) {
+    Player* player = GET_PLAYER(play);
+    s16 yawDiff;
+    s32 absYawDiff;
+    f32 posZ;
+    f32 posX;
+
+    if (this->transformedPlayerPos.z > 0.0f) {
+        return false;
+    }
+
+    yawDiff = this->baseRotY - Camera_GetCamDirYaw(GET_ACTIVE_CAM(play));
+    absYawDiff = ABS_ALT(yawDiff);
+
+    if (absYawDiff < 0x4000) {
+        if (absYawDiff < 0x1800) {
+            this->lightningHitSomething = false;
+        }
+        return false;
+    }
+
+    // If Goht hit the player with lightning the last time it waited, then the only way for `lightningHitSomething` to
+    // be false (in other words, the only way for us continue in this function) is because it was set to false by the
+    // code above in a previous call to this function; this only happens if the difference between Goht's base yaw and
+    // the active cam's yaw is small enough.
+    if (this->lightningHitSomething == true) {
+        return false;
+    }
+
+    posX = this->actor.world.pos.x * this->direction;
+    posZ = this->actor.world.pos.z * this->direction;
+
+    if (((player->actor.world.pos.x > 1200.0f) && (player->actor.world.pos.z < 1200.0f) &&
+         (player->actor.world.pos.z > -1200.0f) && (this->actor.world.pos.x < 0.0f) && (posZ > 1200.0f)) ||
+        ((player->actor.world.pos.x < -1200.0f) && (player->actor.world.pos.z < 1200.0f) &&
+         (player->actor.world.pos.z > -1200.0f) && (this->actor.world.pos.x > 0.0f) && (posZ < -1200.0f)) ||
+        ((player->actor.world.pos.z > 1200.0f) && (player->actor.world.pos.x < 1200.0f) &&
+         (player->actor.world.pos.x > -1200.0f) && (this->actor.world.pos.z < 0.0f) && (posX < -1200.0f)) ||
+        ((player->actor.world.pos.z < -1200.0f) && (player->actor.world.pos.x < 1200.0f) &&
+         (player->actor.world.pos.x > -1200.0f) && (this->actor.world.pos.z > 0.0f) && (posX > 1200.0f))) {
+        return true;
+    }
```

Comment:
The arena looks like this (not to scale):
![image](https://github.com/zeldaret/mm/assets/12245827/cd55fb95-87a6-4549-82ac-9094ff183af9)

The first check (the one before the first `||`) can be visualized like so (again not to scale);
![image](https://github.com/zeldaret/mm/assets/12245827/f249a9ad-0c12-4414-a4b9-48add9c4415b)

Which Goht position is "acceptable" depends on whether Goht is running clockwise or counterclockwise; I tried to use colors to indicate that, but hopefully it makes sense. There are four checks in this if statement because the player can be in any of the four "sides" of the arena excluding the corners; each check is essentially just the same thing, just rotated.

I'm not sure how best to macro-ize this, because the logic is very specific. I guess I could try to write a comment to describe what this is doing? Or just put some ASCII art in there maybe?

Reply:
Sorry, I don't follow. For whatever reason I thought the arena was more "circular", not like the one in the image, so I can't really visualize it.

Reply:
Ahhh, I see. Yeah, the room is actually square with rounded corners. Here are the room DLs from above, minus the ceiling DL:
![image](https://github.com/zeldaret/mm/assets/12245827/29ef205a-dc9d-400e-a664-431ad2b4e3b7)

Does that help with understanding my visualization?

Reply:
And here's a recreation of my original diagram using the room DLs. It isn't perfect, but hopefully putting it in the context of the in-game graphics helps you understand it more:
![image](https://github.com/zeldaret/mm/assets/12245827/13c11cfe-01d1-4a97-94ed-281c6b1edf4c)

Reply:
Ohhh, now I see it! Thanks for the extra images

I think a comment explaining what's going on would be nice to have

Reply:
I just pushed an update with an explanatory comment. It even has this diagram as ASCII art. Let me know what you think.